src/Processors/QueryPlan/JoinStep.cpp

```diff
@@ -0,0 +1,89 @@
+#include <Processors/QueryPlan/JoinStep.h>
+#include <Processors/QueryPipeline.h>
+#include <Processors/Transforms/JoiningTransform.h>
+#include <Interpreters/IJoin.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+JoinStep::JoinStep(
+    const DataStream & left_stream_,
+    const DataStream & right_stream_,
+    JoinPtr join_,
+    size_t max_block_size_)
+    : join(std::move(join_))
+    , max_block_size(max_block_size_)
+{
+    input_streams = {left_stream_, right_stream_};
+    output_stream = DataStream
+    {
+        .header = JoiningTransform::transformHeader(left_stream_.header, join),
+    };
+}
+
+QueryPipelinePtr JoinStep::updatePipeline(QueryPipelines pipelines, const BuildQueryPipelineSettings &)
```

Comment:
What difference between `updatePipeline` and `transformPipeline`? Why JoinStep uses update and FilledJoinStep uses transform?

Reply:
`updatePipeline` is implemented via `transformPipeline` for `ITransformingStep`.